"message" : "Remote URL '' is not valid" when using 'artifactory_remote_repository'

Modifying `artifactory_remote_repository` on version 2.3.10 throws the following error

```
Error: 400 POST https://artifactory.example.com/artifactory/api/repositories/github-remote
{
  "errors" : [ {
    "status" : 400,
    "message" : "Remote URL '' is not valid"
  } ]
}
``` 

```
locals {
  allowed_github_repos = [
    "quixoten/gotee/releases/download/v*/gotee-*",
    "nats-io/gnatsd/releases/download/v*/gnatsd-*"
  ]
}
resource "artifactory_remote_repository" "github-remote" {
  key             = "github-remote"
  package_type    = "generic"
  url             = "https://github.com"
  repo_layout_ref = "simple-default"
  notes           = "managed by terraform"
  content_synchronisation {
    enabled = false
  }
  bypass_head_requests = true
  property_sets = [
    "artifactory"
  ]
  includes_pattern = join(", ", local.allowed_github_repos)
}
```

Here is the tracelog

```
artifactory_remote_repository.github-remote: Modifying... [id=github-remote]
2021/09/22 13:54:03 [DEBUG] EvalApply: ProviderMeta config value set
2021/09/22 13:54:03 [DEBUG] artifactory_remote_repository.github-remote: applying the planned Update change
2021/09/22 13:54:03 [TRACE] GRPCProvider: ApplyResourceChange
2021/09/22 13:54:03 [DEBUG] artifactory_remote_repository.github-remote: apply errored, but we're indicating that via the Error pointer rather than returning it: 400 POST https://artifactory.example.com/artifactory/api/repositories/github-remote
{
  "errors" : [ {
    "status" : 400,
    "message" : "Remote URL '' is not valid"
  } ]
}
2021-09-22T13:54:03.768-0600 [TRACE] plugin.stdio: received data: channel=STDERR len=226
2021-09-22T13:54:03.768-0600 [TRACE] plugin.stdio: waiting for stdio data
```

I'm not sure if this is a terraform or a artifactory bug. Looking at the artifactory logs it seems like terraform is passing bad data that is causing an error in artifactory
```
2021-09-22T20:01:46.333Z [jfrt ] [ERROR] [ec40edc6a5c264a6] [o.a.a.r.RestAddonImpl:1306    ] [tp-nio-8081-exec-274] - Unable to update repository: Remote URL '' is not valid
java.lang.IllegalArgumentException: Remote URL '' is not valid
	at org.artifactory.addon.repository.RemoteRepositoryCreateUpdateHelper.replaceOrUpdate(RemoteRepositoryCreateUpdateHelper.java:101)
	at org.artifactory.addon.rest.RestAddonImpl.updateRepository(RestAddonImpl.java:1257)
	at jdk.internal.reflect.GeneratedMethodAccessor2563.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:344)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
	at com.sun.proxy.$Proxy406.updateRepository(Unknown Source)
	at org.artifactory.rest.resource.repositories.RepositoriesResource.updateRepository(RepositoriesResource.java:167)
	at jdk.internal.reflect.GeneratedMethodAccessor2562.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:475)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:397)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:366)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:319)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.artifactory.webapp.servlet.RepoFilter.execute(RepoFilter.java:203)
	at org.artifactory.webapp.servlet.RepoFilter.doFilter(RepoFilter.java:105)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.artifactory.webapp.servlet.AccessFilter.useAuthentication(AccessFilter.java:563)
	at org.artifactory.webapp.servlet.AccessFilter.authenticateAndExecute(AccessFilter.java:419)
	at org.artifactory.webapp.servlet.AccessFilter.doFilterInternal(AccessFilter.java:268)
	at org.artifactory.webapp.servlet.AccessFilter.doFilter(AccessFilter.java:207)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.artifactory.webapp.servlet.RequestFilter.doFilter(RequestFilter.java:87)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.artifactory.webapp.servlet.ArtifactoryCsrfFilter.doFilter(ArtifactoryCsrfFilter.java:96)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.springframework.session.web.http.SessionRepositoryFilter.doFilterInternal(SessionRepositoryFilter.java:164)
	at org.springframework.session.web.http.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:80)
	at org.artifactory.webapp.servlet.SessionFilter.doFilter(SessionFilter.java:67)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.artifactory.webapp.servlet.ArtifactoryTracingFilter.doFilter(ArtifactoryTracingFilter.java:32)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.artifactory.webapp.servlet.ArtifactoryFilter.doFilter(ArtifactoryFilter.java:123)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:201)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:544)
	at org.apache.catalina.valves.rewrite.RewriteValve.invoke(RewriteValve.java:305)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:143)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:81)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:364)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:616)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:831)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1629)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: java.net.MalformedURLException: no protocol: 
	at java.base/java.net.URL.<init>(URL.java:645)
	at java.base/java.net.URL.<init>(URL.java:541)
	at java.base/java.net.URL.<init>(URL.java:488)
	at org.artifactory.addon.repository.RemoteRepositoryCreateUpdateHelper.replaceOrUpdate(RemoteRepositoryCreateUpdateHelper.java:99)
	... 79 common frames omitted

```

Looking at the artifactory API, I don't see `Remote URL` as a parameter to any api. I'm happy to get more debugging information since I can reproduce this every terraform run. I've used `TF_LOG=TRACE` but it doesn't have any relevant information. Any guidance on how to further debug this appreciated.

--- a/pkg/artifactory/resource_artifactory_remote_repository.go
+++ b/pkg/artifactory/resource_artifactory_remote_repository.go
@@ -360,7 +360,7 @@ func unpackRemoteRepo(s *schema.ResourceData) (MessyRemoteRepo, error) {
 	repo.SuppressPomConsistencyChecks = d.getBoolRef("suppress_pom_consistency_checks", true)
 	repo.SynchronizeProperties = d.getBoolRef("synchronize_properties", true)
 	repo.UnusedArtifactsCleanupPeriodHours = d.getInt("unused_artifacts_cleanup_period_hours", true)
-	repo.Url = d.getString("url", true)
+	repo.Url = d.getString("url", false)
 	repo.Username = d.getString("username", true)
 	repo.VcsGitDownloadUrl = d.getString("vcs_git_download_url", true)
 	repo.VcsGitProvider = d.getString("vcs_git_provider", true)

--- a/pkg/artifactory/resource_artifactory_remote_repository_test.go
+++ b/pkg/artifactory/resource_artifactory_remote_repository_test.go
@@ -62,6 +62,82 @@ func TestKeyHasSpecialCharsFails(t *testing.T) {
 	})
 }
 
+func TestAccRemoteRepositoryChangeConfigGH148(t *testing.T) {
+	_,fqrn,name := mkNames("github-remote","artifactory_remote_repository")
+	const step1 = `
+		locals {
+		  allowed_github_repos = [
+			"quixoten/gotee/releases/download/v*/gotee-*",
+			"nats-io/gnatsd/releases/download/v*/gnatsd-*"
+		  ]
+		}
+		resource "artifactory_remote_repository" "{{ .name }}" {
+		  key = "{{ .name }}"
+		  package_type = "generic"
+		  url = "https://github.com"
+		  repo_layout_ref = "simple-default"
+		  notes = "managed by terraform"
+		  content_synchronisation {
+			enabled = false
+		  }
+		  bypass_head_requests = true
+		  property_sets = [
+			"artifactory"
+		  ]
+		  includes_pattern = join(", ", local.allowed_github_repos)
+		}
+	`
+	const step2 = `
+		locals {
+		  allowed_github_repos = [
+			"quixoten/gotee/releases/download/v*/gotee-*"
+		  ]
+		}
+		resource "artifactory_remote_repository" "{{ .name }}" {
+		  key = "{{ .name }}"
+		  package_type = "generic"
+		  url = "https://github.com"
+		  repo_layout_ref = "simple-default"
+		  notes = "managed by terraform"
+		  content_synchronisation {
+			enabled = false
+		  }
+		  bypass_head_requests = true
+		  property_sets = [
+			"artifactory"
+		  ]
+		  includes_pattern = join(", ", local.allowed_github_repos)
+		}
+	`
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		CheckDestroy: testAccCheckRepositoryDestroy(fqrn),
+		Providers:    testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: executeTemplate("one", step1, map[string]interface{}{
+					"name" : name,
+				}),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(fqrn, "key", name),
+					resource.TestCheckResourceAttr(fqrn, "package_type", "generic"),
+					resource.TestCheckResourceAttr(fqrn, "url", "https://github.com"),
+				),
+			},
+			{
+				Config: executeTemplate("two", step2, map[string]interface{}{
+					"name" : name,
+				}),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(fqrn, "key", name),
+					resource.TestCheckResourceAttr(fqrn, "package_type", "generic"),
+					resource.TestCheckResourceAttr(fqrn, "url", "https://github.com"),
+				),
+			},
+		},
+	})
+}
+
 func TestAccRemoteRepository_basic(t *testing.T) {
 	id := rand.Int()
 	name := fmt.Sprintf("terraform-remote-test-repo-basic%d", id)

